### PR TITLE
Add completions for all options that take values of `git commit`

### PR DESCRIPTION
**Pull request by hexium310**
# Fix
The completion was not correctly shown when there was a query between `--fixup=` and `**`.
This fixed completion to be shown as below.
```sh
$ git commit --fixup=master**
> --fixup=master
  2/100
> --fixup=master         Add foo bar
  --fixup=origin/master  Add foo bar
```

# Add
- Taking a commit hash
    - --reuse-message/-C
    - --reedit-message/-c
    - --fixup
    - --squash
- Taking a commit message
    - --message/-m
- Taking a file
    - --file/-F
    - --template/-t
- Taking an own value
    - --cleanup
    - --untracked-file/-u
- ~~Taking a GPG key~~
    - ~~--gpg-sign/-S~~

Two options below don't show completions, because it is very difficult deciding the candidate for these value.
- --author
- --date

### CODE REVIEW

File: git.zsh [hunk 1]
```diff
@@ -35,11 +37,85 @@ _fzf_complete_git() {
     fi
 
     if [[ "$@" = 'git commit'* ]]; then
-        if [[ "$prefix" = '--fixup=' ]]; then
+        if [[ "$prefix" =~ '^--(reuse-message|reedit-message|fixup|squash)=' ]]; then
+            prefix_option="${prefix/=*/=}"
+            _fzf_complete_git-commits '' "$@"
+            unset prefix_option
+            return
```
Comment (chitoku-k) [inline git.zsh]: This should work:
```zsh
            prefix_option="${prefix/=*/=}" _fzf_complete_git-commits '' "$@"
            return
```

File: git.zsh [hunk 2]
```diff
@@ -35,11 +37,85 @@ _fzf_complete_git() {
...
+        fi
+
+        if [[ "$prefix" =~ '^--author=' ]]; then
+            return
+        fi
+
+        if [[ "$prefix" =~ '^--date=' ]]; then
+            return
+        fi
+
```
Comment (chitoku-k) [inline git.zsh]: `--author 'Foo Bar <foobar@example.com>'` or `--date 'Mon, 13 Jan 2020 20:42:53 +0900'` should have no completion as well.

File: git.zsh [hunk 3]
```diff
@@ -35,11 +37,85 @@ _fzf_complete_git() {
...
+            _fzf_path_completion '' "$@$prefix"
+            return
+        fi
+
+        if [[ "$last_options" =~ '^(-[^-]*[Ft]|--(file|template))' ]]; then
+            _fzf_path_completion '' "$@"
+            return
+        fi
+
+        local cleanup_mode="strip whitespace verbatim scissors default"
```
Comment (chitoku-k) [inline git.zsh]: Make this array benefits later!
```suggestion
        local cleanup_mode=(strip whitespace verbatim scissors default)
```

File: git.zsh [hunk 4]
```diff
@@ -35,11 +37,85 @@ _fzf_complete_git() {
...
+        if [[ "$prefix" =~ '^--cleanup=' ]]; then
+            _fzf_complete '' "$@" < <(echo $cleanup_mode | awk -v RS=' ' -v prefix="${prefix/=*/=}" '{ print prefix $0 }')
```
Comment (chitoku-k) [inline git.zsh]:
```suggestion
            _fzf_complete '' "$@" < <(awk -v prefix="${prefix/=*/=}" '{ print prefix $0 }' <<< ${(F)cleanup_mode})
```

File: git.zsh [hunk 5]
```diff
@@ -35,11 +37,85 @@ _fzf_complete_git() {
...
+            return
+        fi
+
+        if [[ "$last_options" = '--cleanup' ]]; then
+            _fzf_complete '' "$@" < <(echo $cleanup_mode | tr ' ' '\n')
```
Comment (chitoku-k) [inline git.zsh]:
```suggestion
            _fzf_complete '' "$@" <<< ${(F)cleanp_mode}
```

File: git.zsh [hunk 6]
```diff
@@ -72,14 +148,26 @@ _fzf_complete_git-commits_post() {
...
     shift
-    _fzf_complete "--ansi --tiebreak=index $fzf_options" "$@" < <(git log --color=always --format='%C(yellow)%h%C(reset)  %s' 2> /dev/null)
+    _fzf_complete "--ansi --tiebreak=index $fzf_options" "$@" < <(
+        git log --format='%h %s' 2> /dev/null |
+        awk -v prefix="$prefix_option" '
+            {
+                match($0, / /)
+                print $1, prefix substr($0, RSTART + RLENGTH)
+            }
+        ' |
```
Comment (chitoku-k) [inline git.zsh]:
```suggestion
        ' | _fzf_complete_git_tabularize
```

File: git.zsh [hunk 7]
```diff
@@ -35,11 +37,85 @@ _fzf_complete_git() {
     fi
 
     if [[ "$@" = 'git commit'* ]]; then
-        if [[ "$prefix" = '--fixup=' ]]; then
+        if [[ "$prefix" =~ '^--(reuse-message|reedit-message|fixup|squash)=' ]]; then
```
Comment (chitoku-k) [inline git.zsh]: I want this to be alphabetically sorted:
```suggestion
        if [[ "$prefix" =~ '^--(fixup|reedit-message|reuse-message|squash)=' ]]; then
```

File: git.zsh [hunk 8]
```diff
@@ -107,6 +195,50 @@ _fzf_complete_git-unstaged-files_post() {
...
+                        message = message " " "[Git Local]"
+                    }
+                    print prefix keys[j], types[j], uid message
+                }
+                delete keys
+                delete types
+            }
+        ' | _fzf_complete_git_tabularize
+    )
+}
```
Comment (chitoku-k) [inline git.zsh]: This implementation has weird behaviour in my environment:

```zsh
$ git commit -S **
>
  4/4
> 0000000000000000  [SC] unknown] Chitoku
  0000000000000000  [E] unknown] Chitoku
  0000000000000000  [SC] Chitoku <odango@chitoku.jp>
  0000000000000000  [E] Chitoku <odango@chitoku.jp>
```

```zsh
$ LANG=C gpg2 --list-secret-keys --keyid-format LONG 2> /dev/null
/home/chitoku/.gnupg/pubring.kbx
--------------------------------
sec   rsa4096/0000000000000000 2016-05-17 [SC]
      0000000000000000000000000000000000000000
uid                 [ unknown] Chitoku <odango@chitoku.jp>
ssb   rsa4096/0000000000000000 2016-05-17 [E] [expires: 2020-05-17]

sec   rsa4096/0000000000000000 2020-01-12 [SC]
      0000000000000000000000000000000000000000
uid                 [ultimate] Chitoku <odango@chitoku.jp>
ssb   rsa4096/0000000000000000 2020-01-12 [E]
```

Can you dig into what happens?

File: git.zsh [hunk 9]
```diff
@@ -107,6 +195,50 @@ _fzf_complete_git-unstaged-files_post() {
...
 }
 
+_fzf_complete_git-gpg-key() {
+    local fzf_options="$1"
+    shift
+    local global_key=$(git config --global user.signingkey)
+    local local_key=$(git config --local user.signingkey)
+    _fzf_complete "--ansi $fzf_options" "$@" < <(
+        [[ -n "$gpg_command" ]] && LANG=C "$gpg_command" --list-secret-keys --keyid-format LONG 2> /dev/null |
+        awk -v global_key="$global_key" -v local_key=$local_key -v prefix="$prefix_option" '
```
Comment (chitoku-k) [inline git.zsh]:
```zsh
        if [[ -n "$gpg_command" ]]; then
            LANG=C "$gpg_command" --list-secret-keys --keyid-format LONG 2> /dev/null | awk \
                -v global_key="$global_key" \
                -v local_key=$local_key \
                -v prefix="$prefix_option" '
                    # ( ˘ω˘ )
                ' | _fzf_complete_git_tabularize
        fi
```

File: git.zsh [hunk 10]
```diff
@@ -72,14 +148,26 @@ _fzf_complete_git-commits_post() {
...
-        $1 = ""
-        sub(/^ /, "", $0)
-        print "'\''" $0 "'\''"
+    awk -v prefix="$prefix_option" '{
+        match($0, /  /)
+        str = substr($0, RSTART + RLENGTH)
+        match(str, prefix)
+        str = substr(str, RSTART + RLENGTH)
+        gsub("'\''", "'\''\\'\'''\''", str)
+        print prefix "'\''" str "'\''"
```
Comment (chitoku-k) [inline git.zsh]: This should be a convenient function that can be generally used in other awk invocations. Can you add it to `_fzf_complete_awk_functions `?

File: git.zsh [hunk 11]
```diff
@@ -35,11 +37,85 @@ _fzf_complete_git() {
...
+            _fzf_complete '' "$@" < <(echo $cleanup_mode | awk -v RS=' ' -v prefix="${prefix/=*/=}" '{ print prefix $0 }')
+            return
+        fi
+
+        if [[ "$last_options" = '--cleanup' ]]; then
+            _fzf_complete '' "$@" < <(echo $cleanup_mode | tr ' ' '\n')
+            return
         fi
+
+        local untracked_file_mode="no normal all"
```
Comment (hexium310) [inline git.zsh]:
```suggestion
        local untracked_file_mode=(no normal all)
```

File: git.zsh [hunk 12]
```diff
@@ -35,11 +37,85 @@ _fzf_complete_git() {
...
+        if [[ "$prefix" =~ '^--untracked-files=' ]]; then
+            _fzf_complete '' "$@" < <(echo $untracked_file_mode | awk -v RS=' ' -v prefix="${prefix/=*/=}" '{ print prefix $0 }')
```
Comment (hexium310) [inline git.zsh]:
```suggestion
            _fzf_complete '' "$@" < <(awk -v prefix="${prefix/=*/=}" '{ print prefix $0 }' <<< ${(F)untracked_file_mode})
```

File: git.zsh [hunk 13]
```diff
@@ -35,11 +37,85 @@ _fzf_complete_git() {
...
+            return
+        fi
+
+        if [[ "$last_options" =~ '^(-[^-]*u|--untracked-files)' ]]; then
+            _fzf_complete '' "$@" < <(echo $untracked_file_mode | tr ' ' '\n')
```
Comment (hexium310) [inline git.zsh]:
```suggestion
            _fzf_complete '' "$@" <<< ${(F)untracked_file_mode}
```

File: git.zsh [hunk 14]
```diff
@@ -16,6 +16,14 @@ _fzf_complete_awk_functions='
 
         return sprintf("%s%s%s%s%s%s %s", index_status_color, index_status, reset, work_tree_status_color, work_tree_status, reset, substr($0, 4))
     }
+    function get_after_prefix(str) {
+        match(str, prefix)
+        return substr(str, RSTART + RLENGTH)
+    }
+    function enclose_in_single_quote(str) {
```
Comment (chitoku-k) [inline git.zsh]:
```suggestion
    function quote_by_single_quotations(str) {
```

**Pull request merged**